Feature/inline tags for all

p3r7 opened the pull request:
PR to answer enhancement reported in issue https://github.com/beberlei/metrics/issues/53

Implementation details:
- as all TaggableCollectors are in fact "InlineTaggable", I renamed the interface InlineTaggableGaugeableCollector into TaggableGaugeableCollector
- I added inline tags support for: Telegraf, InfluxDB and Prometheus
- I added global tags support for: DogStatsD

Some stuff to look for before merging:
- I wrote some new UT for InfluxDB and Prometheus but only tested the InfluxDB ones, due to being unable to download the prometheus client lib on my MS Windows station.
- I made the global tags overridable by passing inline tags with similar keys. This is debatable.

Enjoy!

Commit:
Made all taggable collectors inline-taggable

Commit:
Made all taggable collectors inline-taggable

Commit:
Corrected interface name

Commit:
Corrected interface name

Commit:
Added global tags support

Commit:
Added global tags support

Commit:
Corrected phpunit method matchers

Commit:
Corrected phpunit method matchers

Commit:
Corrected my stupid logic

Commit:
Corrected my stupid logic

Commit:
Potential correction UT

Commit:
Potential correction UT

p3r7 commented:
OK, just discovered those automated travis UT checks.

Nevertheless, I'ms struggling to debug my Prometheus UT, due to the aforementioned reasons  :anguished:

lyrixx commented on src/Beberlei/Metrics/Collector/InlineTaggableGaugeableCollector.php:
```diff
@@ -1,64 +0,0 @@
-/**
- * InlineTaggableGaugeableCollector interface.
- */
-interface InlineTaggableGaugeableCollector
```
You can not remove an interface, that's a BC break.

I did not look at the code for now, but If you have to do it, you have to depreciate it and provide an upgrade path.

But honestly, I prefer to not remove it to not break the BC.

baartosz committed:
Added in memory collector

lyrixx committed:
feature #56 Added in memory collector (baartosz)

This PR was squashed before being merged into the 2.x-dev branch (closes #56).

Discussion
----------

Added in memory collector

Hi.
I needed to measure processes' metrics and retrieve them in same request/run and hook it up with existing storing procedures. It will be useful to stick to same interface later when we will move that outside to graphite or something. I think it may be useful for others too in some cases.

Commits
-------

466ee27 Added in memory collector

lyrixx committed:
Fixed CS

lyrixx committed:
minor #57 Fixed CS (lyrixx)

This PR was merged into the 2.x-dev branch.

Discussion
----------

Fixed CS

Commits
-------

0fb6d89 Fixed CS

buffcode commented:
@lyrixx: I would like to pick up this feature for Prometheus. As Prometheus will switch from `TaggableCollector` to `InlineTaggableGaugeableCollector` and thus lose `setTags()`, what is your preference for an upgrade path?

p3r7 commented:
@buffcode : you do not need to remove the setTags(), it could still be used for global tags, shared by all metrics sent with the Prometheus instance.
See my implementation of it: https://github.com/beberlei/metrics/pull/55/files#diff-04d32a4646baea06bfa885af51292796

I sadly didn't quite got the time to work on this branch lately, and would gladly accept someone taking ownership of just helping, specifically with the prometheus UT. I can't debug them on my M$ Windows station, as the prometheus lib won't download via composer on this OS...

lyrixx commented:
> you do not need to remove the setTags(), it could still be used for global tags, shared by all metrics sent with the Prometheus instance.

:+1: 

This PR needs a rebase BTW.

Commit:
Merge remote-tracking branch 'origin/feature/inline-tags-for-all' into feature/inline-tags-for-all

# Conflicts:
#	src/Beberlei/Metrics/Collector/InlineTaggableGaugeableCollector.php

Commit:
Corrected Prometheus UT

Commit:
Corrected Telegraf Symphony dependency injection UT

p3r7 commented:
:raised_hands: OK, I merged (rebase wasn't happy, probably due to the deleted InlineTaggableGaugeableCollector.php) and finally managed to debug the failing Prometheus UTs.

I added back the `InlineTaggableGaugeableCollector` interface, even though no class implements it no more.

Indeed:
- all classes that were `InlineTaggableGaugeableCollector` are now `TaggableCollector` + `TaggableGaugeableCollector`
- all classes that were `TaggableCollector` still are, but now support inline tags, in addition to global ones

@lyrixx : I would like some guidelines regarding how to mark the `InlineTaggableGaugeableCollector` and what you call providing an upgrade path. Is that a sort of documentation ?

Commit:
Added missing interface for Telegraf

Commit:
Corrected Telegraf tag string construction

Commit:
Deprecated InlineTaggableGaugeableCollector, added it back to implems

p3r7 commented:
OK, I made the `DogStatsD` collector implement back `InlineTaggableGaugeableCollector`, and did the same to `Telegraf` one.
I discovered the `@deprecated` annotation and so deprecated the `InlineTaggableGaugeableCollector` interface.

This way, we should not have any BC break. @lyrixx, are you OK with this approach ?

p3r7 commented:
Any news concerning this PR ?

p3r7 commented:
Ouspy, accidental close

Commit:
Added missing function parameter
